milli/src/search/matching_words.rs

```diff
@@ -161,32 +163,77 @@ fn prefix_damerau_levenshtein(source: &[u8], target: &[u8]) -> (u32, usize) {
         last_row.insert(char_s, row);
     }
 
-    let mut minimum = (u32::max_value(), 0);
-
-    for x in 0..=m {
-        let dist = matrix[(n + 1, x + 1)] as u32;
-        if dist < minimum.0 {
-            minimum = (dist, x)
+    let mut minimum = 2;
+    for x in 0..=n {
+        // let dist = matrix[(x + 1, m + 1)];
+        let min_dist = (0..=m).map(|y| matrix[(x + 1, y + 1)]).min().unwrap();
+        if min_dist <= 2 {
+            minimum = x;
         }
     }
 
-    minimum
+    // everything was done characters wise and now we want to returns a number of bytes
+    source.chars().take(minimum).map(|c| c.len_utf8()).sum()
 }
 
 #[cfg(test)]
 mod tests {
+    use std::str::from_utf8;
+
     use super::*;
     use crate::search::query_tree::{Operation, Query, QueryKind};
     use crate::MatchingWords;
 
     #[test]
-    fn matched_length() {
-        let query = "Levenste";
-        let text = "Levenshtein";
-
-        let (dist, length) = prefix_damerau_levenshtein(query.as_bytes(), text.as_bytes());
-        assert_eq!(dist, 1);
-        assert_eq!(&text[..length], "Levenshte");
+    fn test_bytes_to_highlight() {
+        struct TestBytesToHighlight {
+            query: &'static str,
+            text: &'static str,
+            length: usize,
+        }
+        let tests = [
+            TestBytesToHighlight { query: "bip", text: "bip", length: "bip".len() },
+            TestBytesToHighlight { query: "bip", text: "boup", length: "bip".len() },
+            TestBytesToHighlight {
+                query: "Levenshtein",
+                text: "Levenshtein",
+                length: "Levenshtein".len(),
+            },
+            // we get to the end of our word with only one typo
+            TestBytesToHighlight {
+                query: "Levenste",
+                text: "Levenshtein",
+                length: "Levenste".len(),
+            },
+            // we get our third and last authorized typo right on the last character
+            TestBytesToHighlight {
+                query: "Levenstein",
+                text: "Levenshte",
+                length: "Levenstei".len(),
+            },
+            // we get to the end of our word with only two typos at the beginning
+            TestBytesToHighlight {
+                query: "Bavenshtein",
+                text: "Levenshtein",
+                length: "Bavenshtein".len(),
+            },
+            // Since we calculate a distance char by char we are supposed to have only two mistakes
+            // here. That would've not be the case if we were computing the distance bytes per bytes
+            TestBytesToHighlight { query: "Båve", text: "Chiøt", length: "Bå".len() },
+            TestBytesToHighlight { query: "💪🙂🍤", text: "plouf", length: "💪🙂".len() },
+            TestBytesToHighlight { query: "clôu¿i", text: "bloubi", length: "clôu".len() },
```

Comment:
Would it be possible to also add the example from the three issues of @curquiza?

Reply:
done 🙂